contracts/src/cryptography/eip712.rs

```diff
@@ -0,0 +1,246 @@
+//! [EIP-712](https://eips.ethereum.org/EIPS/eip-712) is a standard for hashing
+//! and signing typed structured data.
+//!
+//! The implementation of the domain separator was designed to be as efficient
+//! as possible while still properly updating the chain id to protect against
+//! replay attacks on an eventual fork of the chain.
+//!
+//! NOTE: This contract implements the version of the encoding known as "v4", as
+//! implemented by the JSON RPC method [`eth_signTypedDataV4`] in MetaMask.
+//!
+//! [`eth_signTypedDataV4`]: https://docs.metamask.io/guide/signing-data.html
+
+use alloc::{string::String, vec::Vec};
+
+use alloy_primitives::{
+    b256, fixed_bytes, keccak256, Address, FixedBytes, B256, U256,
+};
+use alloy_sol_types::{sol, SolType};
+use stylus_sdk::{
+    block, contract,
+    stylus_proc::{external, sol_storage},
+};
+
+use super::message_hash_utils::to_typed_data_hash;
+
+/// `keccak256("EIP712Domain(string name,string version,uint256 chainId,address
+/// verifyingContract)");`
+pub const TYPE_HASH: B256 =
+    b256!("8b73c3c69bb8fe3d512ecc4cf759cc79239f7b179b0ffacaa9a75d522b39400f");
+pub const FIELDS: FixedBytes<1> = fixed_bytes!("15");
+pub const SALT: B256 = B256::ZERO;
```

Comment:
By default we should try to use `alloy`'s types as least as possible to improve consistency and compatibility. This would also improve this code -- a tell sign is that you are having to dereference these variables when they are used.

```suggestion
pub const TYPE_HASH: [u8; 32] =
    hex!("8b73c3c69bb8fe3d512ecc4cf759cc79239f7b179b0ffacaa9a75d522b39400f");
pub const FIELDS: [u8; 2] = hex!("0f");
pub const SALT: [u8; 32] = [0u8; 32];
```